File: Sources/PhotosIngest/IngestConfig.swift
import Foundation

/// Everything one run needs to know, resolved before it starts.
///
/// A value type rather than a pile of parameters so `PhotosCLI` stays what §7 says it is —
/// argument parsing and wiring — and every rule that matters is reachable from a test that
/// never builds a command line.
public struct IngestConfig: Sendable {

    /// The master copy. The run is read-only on it, apart from pulling down albums the
    /// laptop has never owned (§7).
    public var libraryRoot: URL
    /// Holds `sync_state.db` and `shards/`. Derived and deletable: losing it costs one
    /// re-fetch of every shard.
    public var cacheRoot: URL
    /// Where derivatives are staged before upload. Removed when the run ends.
    public var workRoot: URL

    /// Encoder workers. One per core, measured at 0.43 s/photo wall on 16 (§7).
    public var jobs: Int
    /// Upload connections. **One**, because §9 measured 1.9 MB/s on one stream against
    /// 1.5 MB/s on eight — parallel uploads are slower, not faster. The knob is here for a
    /// different link, not for this one.
    public var uploadJobs: Int

    /// Restricts the run to albums whose source path contains this, case-insensitively.
    /// It scopes deletions and pulls as well as uploads: a scoped run that deleted
    /// everything outside its scope would be a trap.
    public var albumFilter: String?
    /// Plan and print, change nothing. The only safety surface there is (§7).
    public var dryRun: Bool

    /// How old an unreferenced blob must be before the sweep may delete it.
    ///
    /// Seven days is not a guess: presigned URLs live at most 7 days (§1) and §8's
    /// background uploads run against them, so a blob older than that cannot belong to an
    /// upload that can still complete. Below it, an unreferenced blob is indistinguishable
    /// from one the phone is uploading right now.
    public var sweepAge: TimeInterval

    public init(
        libraryRoot: URL,
        cacheRoot: URL,
        workRoot: URL,
        jobs: Int = ProcessInfo.processInfo.activeProcessorCount,
        uploadJobs: Int = 1,
        albumFilter: String? = nil,
        dryRun: Bool = false,
        sweepAge: TimeInterval = 7 * 24 * 60 * 60
    ) {
        self.libraryRoot = libraryRoot
        self.cacheRoot = cacheRoot
        self.workRoot = workRoot
        self.jobs = max(1, jobs)
        self.uploadJobs = max(1, uploadJobs)
        self.albumFilter = albumFilter
        self.dryRun = dryRun
        self.sweepAge = sweepAge
    }

    /// The default cache location: `$XDG_CACHE_HOME/photos-cli`, which is also what
    /// systemd's `CacheDirectory=` produces.
    public static func defaultCacheRoot(
        environment: [String: String] = ProcessInfo.processInfo.environment
    ) -> URL {
        if let explicit = environment["PHOTOS_CACHE_DIR"], !explicit.isEmpty {
            return URL(fileURLWithPath: explicit)
        }
        let base = environment["XDG_CACHE_HOME"].flatMap { $0.isEmpty ? nil : $0 }
            .map { URL(fileURLWithPath: $0) }
            ?? FileManager.default.homeDirectoryForCurrentUser.appending(path: ".cache")
        return base.appending(path: "photos-cli")
    }

    /// A fresh staging directory under `$TMPDIR`. One per run, so debris from a run that
    /// died is a directory a tmp-cleaner can see rather than something to reason about.
    public static func newWorkRoot(
        environment: [String: String] = ProcessInfo.processInfo.environment
    ) -> URL {
        let base = environment["TMPDIR"].flatMap { $0.isEmpty ? nil : $0 }
            .map { URL(fileURLWithPath: $0) } ?? URL(fileURLWithPath: "/tmp")
        return base.appending(path: "photos-cli-\(UUID().uuidString.prefix(8))")
    }
}

/// The conditions under which a run refuses to change anything (exit 3).
///
/// All three mean the same thing: the library is not in the state the catalog was built
/// from, and proceeding would write something irreversible on a false premise.
public enum IngestAbort: Error, CustomStringConvertible {

    /// No readable `$LIBRARY_ROOT/.photosignore`.
    ///
    /// The file is the marker that says "this directory is a library root", which is what
    /// makes unattended deletion safe: an unmounted disk is a bare mount point, and a
    /// mistyped root is somebody else's directory — neither has one. A library that wants
    /// no exclusions writes an empty file.
    case notALibraryRoot(URL, detail: String)

    /// A file's size no longer matches the row that describes it. §7 asserts images never
    /// change on disk; a mismatch means the library broke that contract, so nothing is
    /// written and a person is told, rather than the file being silently re-ingested.
    case fileChanged([ByteMismatch])

    /// A shard too new to read whose two stable columns could not be read either, so the
    /// folder it claims is unknowable. Continuing would risk uploading that folder as a
    /// second album (§3).
    case unidentifiableShard(String)

    public var description: String {
        switch self {
        case .notALibraryRoot(let url, let detail):
            "\(url.path) is not a library root: \(detail)"
        case .fileChanged(let mismatches):
            "\(mismatches.count) file(s) changed on disk since they were ingested: "
            + mismatches.prefix(3).map(\.description).joined(separator: "; ")
        case .unidentifiableShard(let detail):
            "a shard could not be identified: \(detail)"
        }
    }
}

/// A file whose size disagrees with the row describing it.
public struct ByteMismatch: Hashable, Sendable, CustomStringConvertible {
    public var albumPath: String
    public var filename: String
    public var recorded: Int64
    public var found: Int64

    public var description: String {
        "\(albumPath)/\(filename): catalog says \(recorded) bytes, disk has \(found)"
    }
}

/// Process exit codes (§7).
///
/// `1` and `3` are both failures to `OnFailure=`, but they say different things to a person
/// reading the journal: `1` means the run finished and the zone changed, `3` means it
/// refused to start and the zone is untouched.
public enum ExitCode {
    public static let clean: Int32 = 0
    public static let completedWithFailures: Int32 = 1
    public static let usage: Int32 = 2
    public static let aborted: Int32 = 3
    /// `EX_TEMPFAIL`. The keyring is locked because nobody has logged in yet — not a
    /// failure, just not now. H's unit sets `SuccessExitStatus=75` so it stays quiet.
    ///
    /// Reachable only on the production path. A development run takes `PHOTOS_PASSWORD`
    /// from the environment and never consults a keyring, so it can never defer.
    public static let deferred: Int32 = 75
}

File: domain/src/commonMain/kotlin/net/stho/photos/ingest/IngestConfig.kt
package net.stho.photos.ingest

import kotlin.time.Duration
import net.stho.photos.catalog.AlbumState
import kotlin.time.Duration.Companion.days
import kotlinx.io.files.Path

/**
 * Everything one run needs to know, resolved before it starts.
 *
 * A value rather than a pile of parameters so the CLI stays what §7 says it is — argument parsing
 * and wiring — and every rule that matters is reachable from a test that never builds a command
 * line.
 *
 * Not a `data class`, because [jobs] and [uploadJobs] are clamped rather than stored as given.
 */
public class IngestConfig(
    /**
     * The master copy. The run is read-only on it, apart from pulling down albums the laptop has
     * never owned (§7).
     */
    public val libraryRoot: Path,
    /**
     * Holds `sync_state.db` and `shards/`. Derived and deletable: losing it costs one re-fetch of
     * every shard.
     *
     * There is no `defaultCacheRoot` here: that is `Paths.cacheRoot`, the port §7 already
     * declares for "where this device keeps things", and its XDG rules belong in the Linux
     * adapter rather than in a second copy under the domain.
     */
    public val cacheRoot: Path,
    /**
     * Encoder workers. One per core, measured at 0.43 s/photo wall on 16 (§7).
     *
     * No default: the core count is a platform question, and reaching for it from the domain is
     * the third of §7's three reasons a seam becomes a port. The CLI passes it.
     */
    jobs: Int,
    /**
     * Upload connections. **One**, because §9 measured 1.9 MB/s on one stream against 1.5 MB/s on
     * eight — parallel uploads are slower, not faster. The knob is here for a different link, not
     * for this one.
     */
    uploadJobs: Int = 1,
    /**
     * Delete connections. **Sixty-four**, and the opposite reasoning to [uploadJobs]: a delete
     * carries no bytes, so it is not competing for the upstream link — it is one round trip to
     * Frankfurt and back, and round trips overlap.
     *
     * Measured against the live zone while emptying it: a single delete costs **~1.4 s**, and
     * 64 in flight sustained **~45/s**. Serially that is 34,000 blobs in about thirteen hours,
     * which is what a profile bump orphans (§5) against a three-hour import. The retry policy
     * covers 429 and 5xx, so a server that dislikes the rate says so and the run backs off.
     */
    deleteJobs: Int = 64,
    /**
     * Restricts the run to albums whose source path contains this, case-insensitively. It scopes
     * deletions and pulls as well as uploads: a scoped run that deleted everything outside its
     * scope would be a trap.
     */
    public val albumFilter: String? = null,
    /** Plan and print, change nothing. The only safety surface there is (§7). */
    public val dryRun: Boolean = false,
    /**
     * How long an album may sit at [AlbumState.UPLOADING] before it counts as abandoned.
     *
     * Seven days is not a guess: presigned URLs live at most 7 days (§1) and §8's background
     * uploads run against them, so past that the upload provably cannot finish.
     *
     * It no longer gates garbage collection. §8 names every blob an upload will write before
     * writing any of them, so an unreferenced blob cannot belong to something in flight and is
     * collected at once however new (§2). Liveness and garbage stopped sharing this knob.
     */
    public val sweepAge: Duration = 7.days,
) {
    public val jobs: Int = jobs.coerceAtLeast(1)
    public val uploadJobs: Int = uploadJobs.coerceAtLeast(1)
    public val deleteJobs: Int = deleteJobs.coerceAtLeast(1)

    /**
     * Where derivatives are staged before upload. Emptied when the run starts and again when it
     * ends.
     *
     * Under [cacheRoot] rather than `$TMPDIR`, and that is the whole fix for a run staging video
     * transcodes in RAM: `/tmp` is tmpfs on a normal Linux desktop, so a 39-hour import used to
     * hold every transcode in memory until its upload returned. It also buys the cleanup rule.
     * The run lock lives in [cacheRoot], so a run that holds it is the only run that can be using
     * this directory — which makes anything found here at startup debris from a run that died,
     * with no pid to check and no age to guess at. `finally` cannot be the only cleanup: `SIGKILL`
     * skips it, and an out-of-memory kill is exactly how a run staging into RAM ends.
     *
     * Derived rather than passed in, because a caller that could point it somewhere else could
     * point two runs at one directory, and the lock that makes emptying it safe would not follow.
     */
    public val workRoot: Path = Path(cacheRoot, "work")
}

/**
 * Process exit codes (§7).
 *
 * `1` and `3` are both failures to `OnFailure=`, but they say different things to a person reading
 * the journal: `1` means the run finished and the zone changed, `3` means it refused to start and
 * the zone is untouched.
 */
public object ExitCode {
    public const val CLEAN: Int = 0
    public const val COMPLETED_WITH_FAILURES: Int = 1
    public const val USAGE: Int = 2
    public const val ABORTED: Int = 3

    /**
     * `EX_TEMPFAIL`. Not now, rather than not working. Two things say it: the keyring is still
     * locked because nobody has logged in yet, and another sync already holds the run lock. The
     * systemd unit sets `SuccessExitStatus=75` so neither reaches `OnFailure=` — both are the
     * ordinary state of a machine that reboots and an import that outlasts the hour between timer
     * firings.
     */
    public const val DEFERRED: Int = 75
}
